### PR TITLE
Unique subsystem of rxn

### PR DESCRIPTION
**Pull request by cheng-yu-zhang**
### Main improvements in this PR:
*Try to be as clear as possible: Is it fixing/adding something in the model? Is it an additional test/function/dataset? PLEASE 
Add subsystem of rxns systematacially and the data is in file ‘Rxn_unique_subsystem.tsv’
The maping file is in 'KEGG_pathways_maping.tsv'.
**I hereby confirm that I have:**

- [x] Tested my code with [all requirements](https://github.com/SysBioChalmers/yeast-GEM#required-software---contributor) for running the model
- [x] Selected `develop` as a target branch (top left drop-down menu)
- [x] If needed, asked first in the [Gitter chat room](https://gitter.im/SysBioChalmers/yeast-GEM) about this PR

### CODE REVIEW

**Comment by edkerk** [comment]
I rebased it to the latest `develop` branch. But it now requires a function to apply these changes.

A few comments:
- [ ] Remove the `sceXXXXX` codes at the end of some subsystems. KEGG pathways can be stored as separate annotation.
- [ ] Simplify `glycerolipid metabolism /  glycerophospholipid metabolism ( sce00561,sce00564 )` into `glycerolipid metabolism`

**Comment by edkerk** [comment]
Will you also write a function to apply those curations? And are the two files not mostly containing the same information?

**Comment by hongzhonglu** [comment]
> Will you also write a function to apply those curations? And are the two files not mostly containing the same information?

Hi Ed, do you mean  find some applications for this curations? The unique subsystem is mainly used to generate the metabolic map. May be also helpful for the function classification of reactions.

**Comment by edkerk** [comment]
No application. But at the moment it is just a text file with subsystems, it's not incorporated in the model. I can also do that, just wondered if this was worked on already, otherwise it would be double work.

**Comment by mihai-sysbio** [comment]
I too am looking forward to having the unique subsystem integrated in the model file(s).

**Comment by hongzhonglu** [comment]
Yes. These unique subsystem should be worked directly.  But in future maybe we need to consider how to organize the subsystem in a more reasonable way as there are different standards-pathway definition in KEGG and biocyc. @edkerk

**Comment by edkerk** [comment]
KEGG pathways are on [identifiers.org](https://registry.identifiers.org/registry/kegg.pathway), so they can easily be incorporated as `kegg.pathway` annotations. Does [`KEGG_pathways_maping.tsv`](https://github.com/SysBioChalmers/yeast-GEM/pull/307/files#diff-12aa2a702dd1c2587973f141a9c87d113c1c0f3d761b94a9ffede192b73cd445) supposed to contain those annotations? It seems to be the right length (`numel(model.rxns)`), but the file doesn't mention reaction IDs, so a bit dangerous to match just by order in the file.

File: data/modelCuration/Rxn_unique_subsystem.tsv [hunk 1]
```diff
@@ -0,0 +1,4059 @@
...
+r_2662	MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:0, 2-16:1) acyltransferase, mitochondrial membrane	monolysocardiolipin (2-16:1, 3-16:1, 4-18:1)[mm] + phosphatidylcholine (1-18:0, 2-16:1)[mm] <=> cardiolipin (1-16:1, 2-16:1, 3-16:1, 4-18:1)[mm] + 1-acylglycerophosphocholine (18:0)[mm]	YPR140W	glycerophospholipid metabolism
+r_2663	MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-18:1, 2-16:1) acyltransferase, mitochondrial membrane	monolysocardiolipin (2-16:1, 3-16:1, 4-18:1)[mm] + phosphatidylcholine (1-18:1, 2-16:1)[mm] <=> cardiolipin (1-16:1, 2-16:1, 3-16:1, 4-18:1)[mm] + 1-acylglycerophosphocholine (18:1)[mm]	YPR140W	glycerophospholipid metabolism
+r_2664	MLCL (2-16:1, 3-16:1, 4-18:1):PC (1-16:0, 2-18:1) acyltransferase, mitochondrial membrane	monolysocardiolipin (2-16:1, 3-16:1, 4-18:1)[mm] + phosphatidylcholine (1-16:0, 2-18:1)[mm] <=> cardiolipin (1-18:1, 2-16:1, 3-16:1, 4-18:1)[mm] + 1-acylglyc
```
Comment (edkerk) [inline data/modelCuration/Rxn_unique_subsystem.tsv:3990]: Here and elsewhere, please write out `FA ester pathway` as `Fatty acid ester pathway`, abbreviations are in principle not used in subSystem names.

File: data/modelCuration/Rxn_unique_subsystem.tsv [hunk 2]
```diff
@@ -0,0 +1,4059 @@
...
+r_0988	saccharopine dehydrogenase (NAD, L-lysine forming)	H2O[c] + L-saccharopine[c] + NAD[c] => 2-oxoglutarate[c] + H+[c] + L-lysine[c] + NADH[c]	YIR034C	lysine metabolism
+r_0989	saccharopine dehydrogenase (NADP, L-glutamate forming)	H+[c] + L-allysine[c] + L-glutamate[c] + NADPH[c] => H2O[c] + L-saccharopine[c] + NADP(+)[c]	YNR050C	lysine metabolism
+r_0990	sedoheptulose 1,7-bisphosphate D-glyceraldehyde-3-phosphate-lyase	sedoheptulose 1,7-bisphosphate[c] <=> D-erythrose 4-phosphate[c] + dihydroxyacetone phosphate[c]	YKL060C	glycolysis / gluconeogenesis
+r_0992	serine O-acetyltransferase	acetyl-CoA[c] + L-serine[c] => coenzyme A[c] + O-acetyl-L-serine[c]	(YDL040C and YGR147C and YHR013C) or (YDL040C and YGR147C)	cysteine and methionine metabolism
+r_0993	serine palmitotransferase	H+[er] + L-serine[er] + palmitoyl-CoA[er] => 3-ketosphinganine[er] + carbon dioxide[er] + coenzyme A[er]	(YBR058C-A and YDR062W and YMR296C) or (YDR062W and YMR296C)	sphingolipid metabolism
+r_0995	seryl-tRNA synthetase	ATP[c] + L-serine[c] + tRNA(Ser)[c] => AMP[c] + diphosphate[c] + Ser-tRNA(Ser)[c]	YDR023W or YHR011W	aminoacyl-trna biosynthesis
```
Comment (edkerk) [inline data/modelCuration/Rxn_unique_subsystem.tsv:763]: Please correctly capitalize `tRNA` in the subSystem name.

File: data/modelCuration/Rxn_unique_subsystem.tsv (hunk 1 above)
Comment (edkerk) [inline data/modelCuration/Rxn_unique_subsystem.tsv:2346]: Capitalize `SLIME`, to clarify that this is an abbreviation and not actual slime (but also no need to write out the abbreviation).

File: data/modelCuration/KEGG_pathways_maping.tsv [hunk 3]
```diff
@@ -0,0 +1,4059 @@
+metabolism	KEGG pathways
+pyruvate metabolism	sce00620
```
Comment (edkerk) [inline data/modelCuration/KEGG_pathways_maping.tsv:2]: Are these KEGG pathways annotations that should be annotated as `kegg.pathway`? Know that annotations can actually have multiple entries, so we can assign reactions with multiple KEGG pathways. Perhaps a strategy to gather this, is to look at the reaction-specific `kegg.reaction` annotations and match those to the KEGG pathways they are involved in?

**Review by edkerk — CHANGES_REQUESTED** [review]
Very valuable work! I just have a few comments to ensure consistency and that most reactions are assigned to a subSystem.

**Comment by edkerk** [comment]
I've also pushed a `v8_6_0` script that applies the change to the model. Note that complete functioning of that script would require #313 to be merged to `develop` and `develop` merged in this branch. But I will not yet do this, as I first want to test how it works when this PR is contributing another change to the same `v8_6_0` script as used in #313. From release 8.6.1 this will all be resolved.

File: data/modelCuration/KEGG_pathways_maping.tsv (hunk 3 above)
Comment (edkerk) [reply data/modelCuration/KEGG_pathways_maping.tsv]: I actually just see that many reactions are already have `kegg.pathway` annotations. Maybe not all, so good still be good to check. Can be in separare Issue/PR though.

Comment (cheng-yu-zhang) [reply data/modelCuration/KEGG_pathways_maping.tsv]: I agree with that. One reaction may belongs to multiple KEGG pathways. And we can solve that in in separare Issue/PR

File: data/modelCuration/Rxn_unique_subsystem.tsv (hunk 1 above)
[the remark this answers is not in the document]
Comment (edkerk) [reply data/modelCuration/Rxn_unique_subsystem.tsv]: Hmm, something went wrong with my comment here. I did not mean those reactions, it was supposed to refer to `r_4046`, which is non-growth associate maintenance, and currently assigned to subsystem "None"

[the remark this answers is not in the document]
Comment (edkerk) [reply data/modelCuration/Rxn_unique_subsystem.tsv]: Will make a separate comment

File: data/modelCuration/Rxn_unique_subsystem.tsv [hunk 4]
```diff
@@ -0,0 +1,4059 @@
...
+r_0090	6-phosphofructo-2-kinase	ATP[c] + D-fructose 6-phosphate[c] => ADP[c] + beta-D-fructose 2,6-bisphosphate[c] + H+[c]	YIL107C or YOL136C	fructose and mannose metabolism
+r_0091	6-phosphogluconolactonase	6-O-phosphono-D-glucono-1,5-lactone[c] + H2O[c] => 6-phospho-D-gluconate[c] + H+[c]	YGR248W or YHR163W	pentose phosphate pathway
+r_0092	6PP-IP5 depyrophosphorylation to IP6	6-diphospho-1D-myo-inositol pentakisphosphate[c] + H2O[c] => H+[c] + myo-inositol hexakisphosphate[c] + phosphate[c]	YLR410W or YOR163W	inositol phosphate metabolism
+r_0093	6PP-IP5 pyrophosphorylation to 5,6-PP2-IP4	6-diphospho-1D-myo-inositol pentakisphosphate[c] + ATP[c] + H+[c] => 5,6-bis(diphospho)-1D-myo-inositol tetrakisphosphate[c] + ADP[c]	YDR017C	phosphatidylinositol signaling system
+r_0094	8-amino-7-oxononanoate synthase	H+[c] + L-alanine[c] + pimeloyl-CoA[c] => 8-amino-7-oxononanoate[c] + carbon dioxide[c] + coenzyme A[c]		biotin metabolism
+r_0095	acetaldehyde condensation	2 acetaldehyde[c] => (R)-acetoin[c]	YGR087C or YLR044C or YLR134W	other
```
[the remark this answers is not in the document]
Comment (edkerk) [reply data/modelCuration/Rxn_unique_subsystem.tsv]: `r_0095` has now been assigned to `Butanoate metabolism`, but others remain. Will make a separate comment.

**Comment by edkerk** [comment]
Keep in mind what the purpose is of `subSystems`, to group reactions that are involved in similar metabolic processes. Often this refers to pathways, but this is not always the case. Having subsystems with just 1 or very few reactions is not very helpful, as it doesn't help with grouping reactions.

Subsystems with very few reactions that might be better merged:
- Merge `Various types of N-glycan biosynthesis` into `N-glycan biosynthesis`.
- One reaction `r_0362` is assigned to `Other types of O-glycan biosynthesis`, but this is a weird name for a subsystem if there are no other subsystems with `O-glycan biosynthesis`. As the reaction is somewhat similar to the reverse of [DPM1](https://www.genome.jp/pathway/sce00510+YPR183W) (making mannan, instead of man-GDP), it can be merged with `N-glycan biosynthesis`.
- Merge `Atrazine degradation`, `Aminobenzoate degradation`, `Styrene degradation`, `Chloroalkane and chloroalkene degradation`, `Limonene and pinene degradation` and `Bisphenol degradation` in `Alternate carbon metabolism`, otherwise they are mostly just 1 reaction per subsystem.

Then there are a few subsystems that are somewhat duplications, which perhaps should be better merged? I'm not 100% certain, but something to consider (@feiranl @mihai-sysbio, what do you think?):
- There are separate `Arginine and proline metabolism` and `Arginine biosynthesis` subsystems, should the latter not be merged with the first?
- Similar with `Phenylalanine metabolism` and `Phenylalanine, tyrosine and tryptophan biosynthesis`?
- Can `Beta-alanine metabolism` be merged with `Alanine, aspartate and glutamate metabolism`?

Finally there are still many reactions with `None` or `Other` as subsystem. This should be avoided.
- Assign `r_0454` fumarate reductase to TCA cycle?
- There are many peptidase reactions, they can be in a subsystem `Dipeptidases`?
- Assign `r_4491` arabinose reductase to `Alternate carbon metabolism`?
- In general, just go through the list of reaction associated to `None` and `Other` and try to assign them to existing subsystems, or define new ones if needed (e.g. `Dipeptidases`, as I cannot think of any other subsystem where they would fit in, but at least they are now all grouped).

Particularly for the last set of changes, it might be good to explicitly document the changes and arguments for each assignment.

**Comment by edkerk** [comment]
@cheng-yu-zhang Why are there two text files, `uniqueSubsystems.tsv` and `keggPathways.tsv`? In `addRxnSubsystem.m`, it uses the `keggPathways.tsv` file to write the subSystems, which seems incorrect, KEGG pathway annotation is something that will be fixed in a separate PR. In addition, there is no need for `addRxnsSubsystem.m`, as `v8_6_0.m` has the required code to make those changes. And this is not a type of curation that will routinely be performed, so there is no need for a generic function.

**Comment by cheng-yu-zhang** [comment]
Most `Other` and `None` have been avoided and the corresponding reasons are provided in `uniqueSubsystem.tsv`, but I can not find proper subsystems for `r_0331`, `r_4198`, `r_4231`.

**Review by edkerk — APPROVED** [review]

**Comment by haowang-bioinfo** [comment]
@cheng-yu-zhang this feature of providing unique subSystem info for reactions is great!

Just a minor comment to `r_2126`, would be nice to adjust its subsystem to `Glycolysis / gluconeogenesis` (use lower case `g` for the first letter of `gluconeogenesis`) which then would be consistent with the others in this group.

File: data/modelCuration/v8_6_0/uniqueSubsystems.tsv [hunk 5]
```diff
@@ -0,0 +1,4059 @@
...
+r_2115	alcohol dehydrogenase, (acetaldehyde to ethanol)	acetaldehyde[c] + H+[c] + NADH[c] => ethanol[c] + NAD[c]	YBR145W or YOL086C	Glycolysis / gluconeogenesis	
+r_2116	acetaldehyde dehydrogenase	acetaldehyde[c] + H2O[c] + NAD[c] => acetate[c] + 2 H+[c] + NADH[c]	YMR110C or YMR170C or YER073W or YOR374W	Glycolysis / gluconeogenesis	
+r_2117	phenylalanine transaminase	L-phenylalanine[c] + pyruvate[c] <=> keto-phenylpyruvate[c] + L-alanine[c]	YHR137W	Phenylalanine, tyrosine and tryptophan biosynthesis	
+r_2118	tryptophan transaminase	keto-phenylpyruvate[c] + L-tryptophan[c] => indole-3-pyruvate[c] + L-phenylalanine[c]	YHR137W	Phenylalanine, tyrosine and tryptophan biosynthesis	
+r_2119	tyrosine transaminase	3-(4-hydroxyphenyl)pyruvate[c] + L-alanine[c] <=> L-tyrosine[c] + pyruvate[c]	YHR137W	Phenylalanine, tyrosine and tryptophan biosynthesis	
+r_2125	coenzyme A: cytoplasm to LP	coenzyme A[c] <=> coenzyme A[lp]		Transport [c, lp]	
+r_2126	sedoheptulose bisphosphatase	H2O[c] + sedoheptulose 1,7-bisphosphate[c] => phosphate[c] + sedoheptulose 7-phosphate[c]	YKR043C	Glycolysis / Gluconeogenesis
```
Comment (haowang-bioinfo) [inline data/modelCuration/v8_6_0/uniqueSubsystems.tsv:1514]:
```suggestion
r_2126	sedoheptulose bisphosphatase	H2O[c] + sedoheptulose 1,7-bisphosphate[c] => phosphate[c] + sedoheptulose 7-phosphate[c]	YKR043C	Glycolysis / gluconeogenesis	
```

Comment (cheng-yu-zhang) [reply data/modelCuration/v8_6_0/uniqueSubsystems.tsv:1514]: I agree with that, that would be nice

**Pull request merged**